unlock-app/src/middlewares/interWindowCommunicationMiddleware.js

```diff
@@ -1,67 +1,105 @@
+            // we need the hash in order to retrieve the account from the iframe URL
+            router.location.pathname + router.location.hash
           )
-          if (storedAddress) {
-            dispatch(setAccount({ address: storedAddress }))
+          if (address) {
+            dispatch(setAccount({ address }))
+            // for subsequent accesses to paywalls, we save the user account in localStorage
+            // this is ONLY a stopgap until we can implement our own paywall-specific provider
+            if (localStorageAvailable(window)) {
+              window.localStorage.setItem('__unlock__account__', address)
+            }
+          } else {
+            if (localStorageAvailable(window)) {
+              // retrieve the stored account saved when the user purchased a key in the past
+              const storedAddress = window.localStorage.getItem(
+                '__unlock__account__'
+              )
+              if (storedAddress) {
+                dispatch(
+                  // the fromLocalStorage flag is needed for the overlay to know
+                  // that it should open a new window, even though the account
+                  // is set.
+                  setAccount({ address: storedAddress, fromLocalStorage: true })
+                )
+              }
+            }
           }
         }
       }
-    }
-    return action => {
-      if (isInIframe && action.type === OPEN_MODAL_IN_NEW_WINDOW) {
-        parent.postMessage('redirect', parent.origin)
-      } else if (!isInIframe && action.type === HIDE_MODAL) {
-        // if the user clicks the button to go to content,
-        // we redirect back to the content, appending as a hash
-        // the user account. This is not sent to the server.
-        // then, the paywall.min.js script detects the hash
-        // and forwards it to the paywall in the iframe.
-        // the code at the top of this file handles that case
-        const {
-          router,
-          account: { address },
-        } = getState()
-        const { redirect } = lockRoute(router.location.pathname)
-        if (redirect) {
-          window.location.href = redirect + '#' + address
+      const { lockAddress, prefix } = lockRoute(router.location.pathname)
+      if (prefix !== 'paywall') return next(action)
+      if (isInIframe) {
+        if (action.type === OPEN_MODAL_IN_NEW_WINDOW) {
+          parent.postMessage('redirect', parent.origin)
+        }
+      } else {
+        if (action.type === HIDE_MODAL) {
+          // if the user clicks the button to go to content,
+          // we redirect back to the content, appending as a hash
+          // the user account. This is not sent to the server.
+          // then, the paywall.min.js script detects the hash
+          // and forwards it to the paywall in the iframe.
+          // the code at the top of this file handles that case
+          redirectToContentFromPaywall(window, getState)
```

Comment:
Can this be repurposed for other Unlock applications later? It seems like it doesn't _need_ to be from the paywall only, right? Perhaps `redirectToContent` or `redirectToContentFromInterstitial`?

Reply:
sure, I don't see why not. Did you want me to rename it? For now, the name is pretty accurate :)